### PR TITLE
General: Create Site Search

### PR DESCRIPTION
**Pull request by RoboticMind**
Creates a basic site search system. Especially useful for the wiki. As the wiki has gotten larger, it has gotten harder to find things on there

Limitations:
* Done client side, so has to send more data and store in memory
* Has to reload page to search again
   * Keeps memory down because it can garbage collect the page data after a search. Already a ~10MB difference in RAM, but that will grow as the site grows.

### CODE REVIEW

**Comment by nathanielcwm** [comment]
Appears like G analytics can track site searches.

https://support.google.com/analytics/answer/1012264?hl=en

**Comment by RoboticMind** [comment]
It would actually still work because this uses `?query=yoursearch` in the URL

I originally planned on making the search work differently through having search bars across the site where using that made more sense. I stuck with it because it would take more work to change and I've already spent too much time on this. As well I may still add search bars to the wiki pages

**Comment by barton2526** [comment]
Could the search bar be made a little wider?  It looks a little squished on my screen:
image

**Comment by barton2526** [comment]
additionally, this is a pretty empty page
image

**Comment by barton2526** [comment]
nit: searching "~" throws an uncaught exception, but we probably don't care

File: assets/js/search.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+  function determineMatchesToShow(matches) {
+      //find how close the matches are and display the first bunch
+      //returns the index of the last match
+
+      if (!matches["content"]) {
+          return null; //doesn't make sense to look if it matched only in the tile
+      }
+
+      const positions = matches["content"].position
```
Comment (barton2526) [inline assets/js/search.js:28]:
```suggestion
      const positions = matches["content"].position;
```

File: assets/js/search.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+      //increment in for loop is done before check, so subtract 1
+      return Math.max(0, i - 1);
+  }
+
+  function closestWordIndex(string, start) {
+      for (var i = start; i > 0 && string[i] != ' '; i--) {
+          //seeks backward until space or start of the string
+      }
+
+      return i
```
Comment (barton2526) [inline assets/js/search.js:46]:
```suggestion
      return i;
```

File: assets/js/search.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      if (results.length) { // display message about no results if there are none
+
+          for (var i = 0; i < results.length; i++) {
+              var result = results[i];
+              var item = store[result.ref];
+
+
+              var li = document.createElement('li');
+
+              var h3 = document.createElement('h3')
```
Comment (barton2526) [inline assets/js/search.js:61]:
```suggestion
              var h3 = document.createElement('h3');
```

File: assets/js/search.js [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+              var a = document.createElement('a');
+              a.href = item.url;
+              a.textContent = item.title;
+
+              h3.appendChild(a)
```
Comment (barton2526) [inline assets/js/search.js:66]:
```suggestion
              h3.appendChild(a);
```

File: assets/js/search.js [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+              li.appendChild(h3);
+
+              var p = document.createElement('p')
```
Comment (barton2526) [inline assets/js/search.js:69]:
```suggestion
              var p = document.createElement('p');
```

File: assets/js/search.js [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+              var matchData = Object.values(result.matchData.metadata)[0]
+              //workaround for weird bug where some search terms don't get the correct label
+
+              var matchCount = determineMatchesToShow(matchData)
```
Comment (barton2526) [inline assets/js/search.js:74]:
```suggestion
              var matchCount = determineMatchesToShow(matchData);
```

File: assets/js/search.js [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+              var a = document.createElement('a');
+              a.href = item.url;
+              a.textContent = item.title;
+
+              h3.appendChild(a)
+              li.appendChild(h3);
+
+              var p = document.createElement('p')
+
+              var matchData = Object.values(result.matchData.metadata)[0]
```
Comment (barton2526) [inline assets/js/search.js:71]:
```suggestion
              var matchData = Object.values(result.matchData.metadata)[0];
```

File: assets/js/search.js [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+                  }
+                  p.textContent += item.content.substring(firstPos, lastPos);
+
+                  if (lastPos != item.content.length - 1) {
+                      p.textContent += "...";
+                  }
+
+              } else {
+                  //if only title or other matches, just put first 150 chars 
+                  p.textContent = item.content.substring(0, 150) + "..."
```
Comment (barton2526) [inline assets/js/search.js:101]:
```suggestion
                  p.textContent = item.content.substring(0, 150) + "...";
```

File: assets/js/search.js [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+  if (searchTerm) {
+      document.getElementById('search-box')
+              .setAttribute("value", searchTerm);
+
+      var idx = lunr(function () {
+          this.field('id');
+          this.field('title', {boost: 10});
+          this.field('content');
+
+          this.metadataWhitelist = ['position']
```
Comment (barton2526) [inline assets/js/search.js:141]:
```suggestion
          this.metadataWhitelist = ['position'];
```

**Comment by barton2526** [comment]
> 
> 
> Could the search bar be made a little wider? It looks a little squished on my screen:
> image

resolved, thanks

**Comment by RoboticMind** [comment]
> additionally, this is a pretty empty page

It's less empty when it's filled with results, but I think with the wider search bar its not as bad.

**Comment by RoboticMind** [comment]
> nit: searching "~" throws an uncaught exception, but we probably don't care

Yeah those are related to extra search parameters you can give with lunr.js. ~ is to try to fuzzy match and is used like `work~1` to say that one character can be off of that word. https://lunrjs.com/guides/searching.html#fuzzy-matches

**Pull request merged**